File: backend/app/repositories/project.py
"""Project model."""

from datetime import datetime
from typing import List, Optional

from beanie import Document
from pydantic import Field


class ProjectMember(Document):
    """Project member schema."""

    user_id: str
    role: str = Field(pattern="^(owner|editor|viewer)$")
    joined_at: datetime = Field(default_factory=datetime.utcnow)


class Project(Document):
    """Project document model."""

    name: str = Field(..., min_length=3, max_length=50)
    subtitle: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    owner_id: str
    members: List[dict] = Field(default_factory=list)  # List of ProjectMember-like dicts
    progress: int = Field(default=0, ge=0, le=100)
    is_template: bool = False
    is_archived: bool = False
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        """Beanie settings."""

        name = "projects"
        indexes = [
            [("owner_id", 1)],
            [("members.user_id", 1)],
            [("is_archived", 1)],
            [("created_at", 1)],  # For sorting projects by creation date
            [("owner_id", 1), ("is_archived", 1)],  # For user's active projects
            [("is_template", 1)],  # For template queries
            [("progress", 1)],  # For progress filtering
        ]

File: frontend/src/types/domain/project.ts
export interface Project {
    id: string
    name: string
    subtitle?: string
    description?: string
    owner_id: string
    members: ProjectMember[]
    progress: number
    is_template: boolean
    is_archived: boolean
    created_at: string
    updated_at: string
}

export interface ProjectMember {
    user_id: string
    role: 'owner' | 'editor' | 'viewer'
    joined_at: string
}

export interface CreateProjectData {
    name: string
    subtitle?: string
    description?: string
    members?: string[]
    is_template?: boolean
}

export interface UpdateProjectData {
    name?: string
    subtitle?: string
    description?: string
    progress?: number
    is_archived?: boolean
}
